src/gui_egui/editor.rs

```diff
@@ -0,0 +1,405 @@
+use crate::common::{Components, EguiComponent, Id, Input};
+use crate::components::*;
+use crate::gui_egui::gui::EguiExtra;
+use crate::gui_egui::{
+    editor_wire_mode::WireMode,
+    gui::Gui,
+    helper::{id_ports_of_all_components_non_wires, offset_helper},
+    keymap,
+    library::InputMode,
+    menu::Menu,
+};
+use eframe::{egui, Frame};
+use egui::{Color32, Context, LayerId, PointerButton, Pos2, Rect, Shape, Vec2};
+use std::{
+    collections::{BTreeMap, HashMap},
+    ops::Range,
+    path::Path,
+    rc::Rc,
+};
+
+pub struct Editor {
+    pub components: Components,
+    pub scale: f32,
+    pub pan: Vec2,
+    pub offset: Vec2,
+    pub offset_and_pan: Vec2,
+    pub clip_rect: Rect,
+    pub side_panel_width: f32,
+    pub ui_change: bool,
+    pub snap_distance: f32,
+    pub grid: GridOptions,
+    pub library: Components,
+    pub dummy_input: Input,
+    pub editor_mode: EditorMode,
+    pub wm: WireMode,
+    pub im: InputMode,
+    pub contexts: HashMap<crate::common::Id, EguiExtra>,
+}
+
+#[derive(Clone)]
+pub struct CloseToComponent {
+    pub comp: Rc<dyn EguiComponent>,
+    pub pos: Pos2,
+    pub dist: f32,
+    pub port_id: Id,
+    pub potential_actual_input: Option<Input>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum EditorMode {
+    Simulator,
+    Default,
+    Wire,
+    Input,
+}
+
+#[cfg(feature = "gui-egui")]
+pub struct EditorRenderReturn {
+    pub delete: bool,
+    pub resp: Option<Vec<egui::Response>>,
+}
+
+// Specific structs for egui
+#[cfg(feature = "gui-egui")]
+pub enum SnapPriority {
+    Default,
+    Wire,
+}
+
+#[derive(Clone)]
+pub struct GridOptions {
+    pub enable: bool,
+    pub size: f32,
+    pub opacity: f32,
+    pub snap_enable: bool,
+    pub snap_distance: f32,
+}
+
+impl Editor {
+    pub fn gui(components: Components, _path: &Path) -> Self {
+        let dummy_input = Input::new("id", "field");
+        let library: Components = vec![
```

Comment:
this should probably be passed as an argument, allowing model crates (like riscv) to provide their own component libraries